File: frontend/src/api/rulesPublic.ts
import api from './client'
import type { RuleDirection } from './rules'

/** T3.11.03 — the public directory. No session, no account: the free half of
 *  stream D (`MASTERPLAN §4.1`). */

export interface PublicSource {
  authority: string
  document_title: string
  document_date: string | null
  url: string
  /** Verbatim. The quotation is what makes the claim checkable. */
  quote: string
}

export interface PublicSection {
  anchor: string
  title: string
  body: string
  /** The locale this section was actually served in — not always the one
   *  asked for. A half-translated corpus is a real state. */
  locale: string
  sources: PublicSource[]
}

export interface PublicQuestion {
  anchor: string
  question: string
  /** Markdown. Short by design — the section behind it carries the citation. */
  answer: string
  /** The section this answer compresses. Rendered as a link down to it: an
   *  answer with nothing behind it would be us asserting somebody else's law
   *  on our own authority. */
  section_anchor: string
  locale: string
}

export interface PublicRequirement {
  code: string
  title: string
  issuer: string
  obtained_by: string
  is_mandatory: boolean
  condition: Record<string, unknown> | null
  valid_for_days: number | null
  /** Days to obtain. The checklist will count this backwards from departure. */
  lead_time_days: number | null
  notes: string
}

export interface PublicRuleSet {
  id: string
  category_key: string
  direction: RuleDirection
  jurisdiction_code: string
  jurisdiction_name: string
  title: string
  version: number
  effective_from: string
  /** A person checked this against the source on this date. */
  reviewed_at: string | null
  /** The watcher last looked at the source. Says nothing about the text. */
  checked_at: string | null
  needs_review: boolean
  /** Something on the page is not in the readers language. */
  fallback_locale: boolean
  locale: string
  /** What the editor said when publishing this version — the "what changed"
   *  line. Empty is the common case for a first version and reads as silence. */
  published_note: string
  /** The compact reading of the same corpus, first on the page. */
  questions: PublicQuestion[]
  sections: PublicSection[]
  requirements: PublicRequirement[]
}

export interface QuestionPreview {
  anchor: string
  question: string
  locale: string
}

export interface RuleIndexEntry {
  category_key: string
  direction: RuleDirection
  jurisdiction_code: string
  jurisdiction_name: string
  title: string
  version: number
  /** Set at publication, so for the directory this is the date the version
   *  went live — what the chronological order sorts by. */
  reviewed_at: string | null
  /** "What changed". Makes the entry read as an entry rather than a menu item;
   *  empty for a first version, which is silence and not a gap. */
  published_note: string
  /** Served by the API, never assembled here: the prerender step writes one
   *  file per path, and a path built in two places differs in one of them. */
  path: string
  /** How many questions this corridor answers. Zero is a real answer. */
  question_count: number
  /** The questions themselves, text only. The directory shows real questions
   *  rather than a count of them: somebody who does not know the taxonomy
   *  still knows their own question. Also what the search field searches. */
  questions: QuestionPreview[]
}

/**
 * T_OPS.2 — the payload the server rendered this page from.
 *
 * Without it the page hydrates against markup full of content while the
 * component starts with no data, React throws the server's work away and
 * re-renders a loading skeleton, and the reader watches a finished page turn
 * into grey boxes and back. Reading the same payload the markup was built from
 * makes the first client render identical to the server one, which is the only
 * thing hydration actually asks for.
 *
 * **The payload carries the address it was rendered for, and that is not
 * belt-and-braces.** The document survives client-side navigation: without the
 * check, opening the catalogue and then clicking into a corridor would hand
 * that page the catalogue's array, which has the wrong shape entirely and would
 * render as garbage rather than fail. Comparing against the current path costs
 * one line and makes the payload usable only where it is true.
 *
 * The id is declared here rather than in `entry-ssr` because both sides need
 * it and only this module is safe to import from the browser bundle.
 *
 * Called by: `pages/RulesIndexPage`, `pages/RulesPage`, `entry-ssr.injectPage`.
 */
export const RULES_DATA_ID = '__rules_data__'

/** `/rules/art/export/RU/` and `/rules/art/export/RU` are the same address. */
const samePath = (a: string, b: string) =>
  a.replace(/\/+$/, '') === b.replace(/\/+$/, '')

export function bootstrapped<T>(): T | undefined {
  if (typeof document === 'undefined') return undefined
  const el = document.getElementById(RULES_DATA_ID)
  if (!el?.textContent) return undefined
  if (!samePath(el.dataset.path || '', window.location.pathname)) return undefined
  try {
    return JSON.parse(el.textContent) as T
  } catch {
    // A malformed payload is not worth a blank page: fall through to the
    // fetch the component would have made anyway.
    return undefined
  }
}

export const rulesIndex = (locale: string) =>
  api.get<RuleIndexEntry[]>('/api/rules', { params: { locale } })

export const readRule = (
  category: string,
  direction: string,
  country: string,
  locale: string,
) =>
  api.get<PublicRuleSet>(`/api/rules/${category}/${direction}/${country}`, {
    params: { locale },
  })

File: backend/app/api/rules_public.py
"""T3.11.03 — the public rules directory. Free, open, and indexable.

No authentication anywhere in this module, and that is the product decision, not
an oversight: `MASTERPLAN §4.1` makes the directory free for the same reason
matching is free — text is copyable, and text nobody can read without an account
is text nobody reads. What is sold is the collected packet, not the knowledge.

Only `published` sets are served. A draft is somebody's work in progress and an
`outdated` one is what a rule used to say; both would read, to a stranger,
exactly like the current answer.

**Every response carries its own trustworthiness.** `reviewed_at` (a person
checked this against the source), `checked_at` (the daily watcher last looked at
the source) and `needs_review` travel as fields, not as a sentence at the bottom
of the page — a machine reading this cannot parse a sentence, and a person
deserves the date next to the claim rather than in a footer.

Endpoints:
- `GET /api/rules` — index of everything published, for the directory and for
  the prerender step that turns it into files.
- `GET /api/rules/{category}/{direction}/{country}` — one set in full.
"""
from __future__ import annotations

import uuid
from datetime import date, datetime

from fastapi import APIRouter, Depends, HTTPException, Query
from pydantic import BaseModel
from sqlalchemy import select
from sqlalchemy.ext.asyncio import AsyncSession

from app.core.database import get_db
from app.models.rules import (
    DocumentRequirement,
    Jurisdiction,
    RuleDirection,
    RuleSection,
    RuleSet,
    RuleSource,
    RuleStatus,
)

router = APIRouter()

#: The corpus is written in these two first; the rest are translated from a
#: settled corpus rather than in parallel with it (IMPLEMENTATIONPLAN §3.11.4).
CORPUS_LOCALES = ("en", "ru")
FALLBACK_LOCALE = "en"


class SourceOut(BaseModel):
    authority: str
    document_title: str
    document_date: date | None
    url: str
    quote: str
    model_config = {"from_attributes": True}


class SectionOut(BaseModel):
    anchor: str
    title: str
    body: str
    #: The locale this section was actually served in. Differs from the one
    #: asked for when the corpus has not been translated yet — the page says so
    #: rather than showing English as if it were the translation.
    locale: str
    sources: list[SourceOut]


class RequirementOut(BaseModel):
    code: str
    title: str
    issuer: str
    obtained_by: str
    is_mandatory: bool
    condition: dict | None
    valid_for_days: int | None
    lead_time_days: int | None
    notes: str
    model_config = {"from_attributes": True}


class RuleIndexOut(BaseModel):
    category_key: str
    direction: RuleDirection
    jurisdiction_code: str
    jurisdiction_name: str
    title: str
    reviewed_at: datetime | None
    #: The address this set is served at. Sent rather than assembled by the
    #: caller: the prerender step writes one file per path, and a path built in
    #: two places is a path that differs in one of them.
    path: str


class RuleSetOut(BaseModel):
    id: uuid.UUID
    category_key: str
    direction: RuleDirection
    jurisdiction_code: str
    jurisdiction_name: str
    title: str
    version: int
    effective_from: datetime
    #: A person checked this text against the source on this date.
    reviewed_at: datetime | None
    #: The watcher last looked at the source on this date. Says nothing about
    #: the text being right — only that we are not blind to the source moving.
    checked_at: datetime | None
    #: Something changed at the source and no one has re-read the rule yet.
    needs_review: bool
    #: True when the reader asked for a locale the corpus does not have yet.
    fallback_locale: bool
    locale: str
    sections: list[SectionOut]
    requirements: list[RequirementOut]


def path_for(category: str, direction: RuleDirection, country: str) -> str:
    """`/rules/<category>/<direction>/<country>/` — category first.

    Owner's decision: people search for "how do I take a painting out of
    Russia", not for "what may leave Russia". The category is the subject of
    the search; the country narrows it.
    """
    return f"/rules/{category}/{direction.value}/{country}/"


@router.get("/rules", response_model=list[RuleIndexOut])
async def index(db: AsyncSession = Depends(get_db)):
    rows = (
        await db.execute(
            select(RuleSet, Jurisdiction.name)
            .join(Jurisdiction, Jurisdiction.code == RuleSet.jurisdiction_code)
            .where(RuleSet.status == RuleStatus.published)
            .order_by(RuleSet.category_key, RuleSet.direction)
        )
    ).all()
    return [
        RuleIndexOut(
            category_key=rs.category_key,
            direction=rs.direction,
            jurisdiction_code=rs.jurisdiction_code,
            jurisdiction_name=name or rs.jurisdiction_code,
            title=rs.title,
            reviewed_at=rs.reviewed_at,
            path=path_for(rs.category_key, rs.direction, rs.jurisdiction_code),
        )
        for rs, name in rows
    ]


@router.get("/rules/{category}/{direction}/{country}", response_model=RuleSetOut)
async def read_rule(
    category: str,
    direction: RuleDirection,
    country: str,
    db: AsyncSession = Depends(get_db),
    locale: str = Query(default=FALLBACK_LOCALE, max_length=5),
):
    rule_set = (
        await db.execute(
            select(RuleSet).where(
                RuleSet.category_key == category,
                RuleSet.direction == direction,
                RuleSet.jurisdiction_code == country.upper(),
                RuleSet.status == RuleStatus.published,
            )
        )
    ).scalar_one_or_none()
    if rule_set is None:
        # 404 for "no such corridor" and for "nothing published here" alike.
        # The distinction is real but not the reader's: an unpublished draft
        # must not be discoverable by the shape of the answer.
        raise HTTPException(status_code=404, detail="No published rules for this corridor")

    wanted = (locale or FALLBACK_LOCALE).split("-")[0].lower()
    if wanted not in CORPUS_LOCALES:
        wanted = FALLBACK_LOCALE

    all_sections = (
        await db.execute(
            select(RuleSection)
            .where(RuleSection.rule_set_id == rule_set.id)
            .order_by(RuleSection.order)
        )
    ).scalars().all()

    # By anchor, preferring the asked-for locale and falling back per section:
    # a corpus half-translated is a real state, and the honest answer is the
    # translated sections in the reader's language and the rest in English —
    # each one saying which it is.
    by_anchor: dict[str, RuleSection] = {}
    for section in all_sections:
        current = by_anchor.get(section.anchor)
        if current is None or (section.locale == wanted and current.locale != wanted):
            by_anchor[section.anchor] = section

    chosen = sorted(by_anchor.values(), key=lambda s: s.order)
    source_rows = (
        await db.execute(
            select(RuleSource).where(
                RuleSource.section_id.in_([s.id for s in chosen] or [uuid.uuid4()])
            )
        )
    ).scalars().all()
    sources: dict[uuid.UUID, list[RuleSource]] = {}
    for src in source_rows:
        sources.setdefault(src.section_id, []).append(src)

    requirements = (
        await db.execute(
            select(DocumentRequirement)
            .where(DocumentRequirement.rule_set_id == rule_set.id)
            .order_by(DocumentRequirement.code)
        )
    ).scalars().all()

    name = (
        await db.execute(
            select(Jurisdiction.name).where(Jurisdiction.code == rule_set.jurisdiction_code)
        )
    ).scalar()

    return RuleSetOut(
        id=rule_set.id,
        category_key=rule_set.category_key,
        direction=rule_set.direction,
        jurisdiction_code=rule_set.jurisdiction_code,
        jurisdiction_name=name or rule_set.jurisdiction_code,
        title=rule_set.title,
        version=rule_set.version,
        effective_from=rule_set.effective_from,
        reviewed_at=rule_set.reviewed_at,
        checked_at=rule_set.checked_at,
        needs_review=rule_set.needs_review,
        # True when *anything* the reader sees is not in their language.
        fallback_locale=any(s.locale != wanted for s in chosen),
        locale=wanted,
        sections=[
            SectionOut(
                anchor=s.anchor,
                title=s.title,
                body=s.body,
                locale=s.locale,
                sources=[SourceOut.model_validate(x) for x in sources.get(s.id, [])],
            )
            for s in chosen
        ],
        requirements=[RequirementOut.model_validate(r) for r in requirements],
    )
